### PR TITLE
plot_contour with target values other than objective value

### PR DESCRIPTION
**Pull request by not522**
## Motivation
This PR enables flexible plotting in `plot_contour`, same as #2064.
## Description of the changes
Added `target` and `target_name` arguments to `plot_contour`.

### CODE REVIEW

File: optuna/visualization/_contour.py [hunk 1]
```diff
@@ -55,17 +61,21 @@ def objective(trial):
...
-            A :class:`~optuna.study.Study` object whose trials are plotted for their objective
-            values.
+            A :class:`~optuna.study.Study` object whose trials are plotted for their target values.
         params:
             Parameter list to visualize. The default is all parameters.
+        target:
+            A function to specify the value to display. If it is :obj:`None`, the objective values
+            are plotted.
+        target_name:
+            Target's name to display on the axis label and the legend.
```
Comment (toshihikoyanase) [inline optuna/visualization/_contour.py:71]: A nit, but `target_name` is only shown in the [color bar](https://plotly.com/python/contour-plots/#color-bar-title) title in case of the contour plots.

image

```suggestion
            Target's name to display on the color bar.
```

File: optuna/visualization/matplotlib/_contour.py [hunk 2]
```diff
@@ -44,10 +50,14 @@ def plot_contour(study: Study, params: Optional[List[str]] = None) -> "Axes":
...
-            A :class:`~optuna.study.Study` object whose trials are plotted for their objective
-            values.
+            A :class:`~optuna.study.Study` object whose trials are plotted for their target values.
         params:
             Parameter list to visualize. The default is all parameters.
+        target:
+            A function to specify the value to display. If it is :obj:`None`, the objective values
+            are plotted.
+        target_name:
+            Target's name to display on the axis label and the legend.
```
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_contour.py:60]: ditto.
```suggestion
            Target's name to display on the color bar.
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. The change basically looks good to me, but I added some small comments.

**Comment by hvy** [comment]
How about making the newly introduced parameter keyword only, as in https://github.com/optuna/optuna/pull/2082 ?

As for the CI failure, it should be fixed if you could merge the latest master branch.

**Comment by not522** [comment]
Thanks, I updated PR following your comments.

File: optuna/visualization/_contour.py [hunk 3]
```diff
@@ -242,13 +266,19 @@ def _generate_contour_subplot(
...
+            value = target(trial)
+
+        if isinstance(value, int):
+            value = float(value)
+        elif not isinstance(value, float):
             raise ValueError(
-                "Trial{} has COMPLETE state, but its value is non-numeric.".format(trial.number)
+                "Trial{} has COMPLETE state, but its target value is non-numeric.".format(
+                    trial.number
+                )
```
Comment (HideakiImamura) [inline optuna/visualization/_contour.py:281]: This is not directly related to this PR change, but what about using f-strings?
```suggestion
                f"Trial{trial.number} has COMPLETE state, but its target value is non-numeric."
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! Basically, LGTM except for a very minor comment. PTAL.

File: optuna/visualization/_contour.py (hunk 3 above)
Comment (not522) [reply optuna/visualization/_contour.py:281]: I also think it's better to use f-strings. Is it recommended in Optuna? Coding Style Conventions does not mention it.

Comment (HideakiImamura) [reply optuna/visualization/_contour.py:281]: According to the issue https://github.com/optuna/optuna/issues/1916#issuecomment-708158413, we'd like to use f-strings when the line doesn't get too long. This is not a strict rule, but now, we can use f-strings without a too long line.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM!
@HideakiImamura Please merge after your review.

**Review by HideakiImamura — APPROVED** [review]
LGTM!
